core/common/src/Segment.kt

```diff
@@ -185,6 +184,84 @@ public class Segment {
         return predecessor
     }
 
+    internal fun writeByte(byte: Byte) {
+        data[limit++] = byte
+    }
+
+    internal fun writeShort(short: Short) {
+        val data = data
+        var limit = limit
+        data[limit++] = (short.toInt() ushr 8 and 0xff).toByte()
+        data[limit++] = (short.toInt() and 0xff).toByte()
+        this.limit = limit
+    }
+
+    internal fun writeInt(int: Int) {
+        val data = data
+        var limit = limit
+        data[limit++] = (int ushr 24 and 0xff).toByte()
+        data[limit++] = (int ushr 16 and 0xff).toByte()
+        data[limit++] = (int ushr 8 and 0xff).toByte()
+        data[limit++] = (int and 0xff).toByte()
+        this.limit = limit
+    }
+
+    internal fun writeLong(long: Long) {
+        val data = data
+        var limit = limit
+        data[limit++] = (long ushr 56 and 0xffL).toByte()
+        data[limit++] = (long ushr 48 and 0xffL).toByte()
+        data[limit++] = (long ushr 40 and 0xffL).toByte()
+        data[limit++] = (long ushr 32 and 0xffL).toByte()
+        data[limit++] = (long ushr 24 and 0xffL).toByte()
+        data[limit++] = (long ushr 16 and 0xffL).toByte()
+        data[limit++] = (long ushr 8 and 0xffL).toByte()
+        data[limit++] = (long and 0xffL).toByte()
+        this.limit = limit
+    }
+
+    internal fun readByte(): Byte {
+        return data[pos++]
+    }
+
+    internal fun readShort(): Short {
+        val data = data
+        var pos = pos
+        val s = (data[pos++] and 0xff shl 8 or (data[pos++] and 0xff)).toShort()
+        this.pos = pos
+        return s
+    }
+
+    internal fun readInt(): Int {
+        val data = data
+        var pos = pos
+        val i = (
+                data[pos++] and 0xff shl 24
+                        or (data[pos++] and 0xff shl 16)
+                        or (data[pos++] and 0xff shl 8)
+                        or (data[pos++] and 0xff)
+                )
+        this.pos = pos
+        return i
+    }
+
+    internal fun readLong(): Long {
+        val data = data
+        var pos = pos
+        val v = (
+                data[pos++] and 0xffL shl 56
+                        or (data[pos++] and 0xffL shl 48)
+                        or (data[pos++] and 0xffL shl 40)
```

Comment:
Maybe it's paranoia, but does it work at the same speed as 
`or (data[pos + 1] and 0xffL shl 40)`
?

(there is no frequent updating of the field value)

Reply:
`pos` here is a local variable defined few lines above, so the field will be updated only once, after all reads are done.

Anyway, I'll re-check how it works with explicit offsets and immutable base, the last time I tested it, there was no significant difference.